### DISCUSSION TITLE
Syntax for active patterns

### DISCUSSION THREAD

**Discussion by WhiteBlackGoose** [dotnet/csharplang#5146] [Language Ideas]
The idea is to use syntax similar to deconstruction *and* extension methods:

```cs
public static class Patterns
{
    public static bool Int(this string t, out int res)
        => int.TryParse(t, out res);
}
```
=>
```cs
var a = "123" switch
{
    { Int: var res } => res,
    _ => throw ...
};
```
or
```cs
if ("123" is { Int: var res })
    ...
```

I don't think there's any need for a keyword, an extension method with `bool` return type and all `out` parameters would be recognized as an active pattern. 

I also don't think there's much of a need for closed active patterns (after all, you can return a DU as an `out` result and switch over it).

More examples:

### Example 1

```cs
public static class Patterns
{
    public static bool Positive(this int i) => i > 0;
}
...
return a switch
{
    { Positive } => "it's positive",
    _ => "it's otherwise"
}
```

### Example 2

```cs
public static class Patterns
{
    public static bool Valid(this SomeObject o) => ...
}
...
someObject switch
{
    { Valid, Id: > 0 } => ...
    { Valid, Id: < 0 } => ...
    { Valid } => ...
    _ => ...
}
```

### Example 3

```cs
public static class PatternsForHospital
{
    public static bool Blind(this Person person) => ...
    public static bool SickOfCancer(this Person person, out CancerKind cancerKind) => ...
    public static bool Dead(this Person person) => ...
}
...
person switch
{
    { Blind, Dead } => "Person is blind and dead",
    { SickOfCancer: var kind, Dead } => $"This dead person had {kind}",
    { SickOfCancer: var kind } => $"This person had {kind}",
    { Blind, Age: > 32 } => $"This one is blind but older than 32" // Age is a property of a person, not pattern. Though syntax is the same
    _ => ...
};
```

**Comment by HaloFour** [comment]
See: https://github.com/dotnet/csharplang/issues/277

**Reply by WhiteBlackGoose** [reply]
Interesting. Though I do not propose the `Is`-pattern naming, and I think it's slightly different, isn't it? Also it was years ago, mb something has changed?

**Reply by HaloFour** [reply]
Yes, the syntax is different but the functionality is more or less the same. I think there are a couple of open proposals for active/user-defined patterns.  Here's another one that was championed by an ex-member of the language team: https://github.com/dotnet/csharplang/issues/1047

My one comment about your specific syntax suggestion is that it is ambiguous with existing recursive property patterns, in that `foo is { Bar: var baz }` is already legal syntax.

**Reply by WhiteBlackGoose** [reply]
That's the point. We're not creating new syntax (except for paramless patterns, it won't require `:`), they will look like properties. I think it's a good idea not to create new syntactic entities for no need

**Reply by HaloFour** [reply]
I wonder if that would fall out naturally from extension properties.

IMO I'd prefer active patterns take a form more like DUs, which would be more likely based on type patterns than property patterns.
